### PR TITLE
Add StronglyTypedResourceBuilder

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Ports StronglyTypedResourceBuilder from .NET Framework.

Note: still adding addtional tests.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7476)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 1]
```diff
@@ -1,5 +1,64 @@
...
+    mimetype: application/x-microsoft.net.object.soap.base64
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
+            : and then encoded with base64 encoding.
+
+    mimetype: application/x-microsoft.net.object.bytearray.base64
+    value   : The object must be serialized into a byte array 
+            : using a System.ComponentModel.TypeConverter
+            : and then encoded with base64 encoding.
+    -->
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:61]: Please remove this comment

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 2]
```diff
@@ -54,9 +54,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\..\System.Windows.Forms.PrivateSourceGenerators\src\System.Windows.Forms.PrivateSourceGenerators.csproj" 
-                      ReferenceOutputAssembly="false"
-                      OutputItemType="Analyzer" />
+    <ProjectReference Include="..\..\System.Windows.Forms.PrivateSourceGenerators\src\System.Windows.Forms.PrivateSourceGenerators.csproj" ReferenceOutputAssembly="false" OutputItemType="Analyzer" />
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj:57]: Please revert this

File: src/System.Windows.Forms.Design/tests/UnitTests/System.Windows.Forms.Design.Tests.csproj [hunk 3]
```diff
@@ -9,6 +9,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.2.0" />
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/System.Windows.Forms.Design.Tests.csproj:12]: Versions must be specified in https://github.com/dotnet/winforms/blob/main/eng/Versions.props

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj [hunk 4]
```diff
@@ -23,9 +23,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\..\System.Windows.Forms.PrivateSourceGenerators\src\System.Windows.Forms.PrivateSourceGenerators.csproj" 
-                      ReferenceOutputAssembly="false"
-                      OutputItemType="Analyzer" />
+    <ProjectReference Include="..\..\System.Windows.Forms.PrivateSourceGenerators\src\System.Windows.Forms.PrivateSourceGenerators.csproj" ReferenceOutputAssembly="false" OutputItemType="Analyzer" />
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System.Windows.Forms.csproj:26]: Please revert

File: src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+            </resheader>
+            <resheader name="writer">
+                <value>System.Resources.ResXResourceWriter</value>
+            </resheader>
+            {data ?? string.Empty}
+        </root>
+        """;
+
+    [Fact]
+    public void Create_NullBaseName_ThrowsArgumentNull()
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs:42]: Please follow the test naming convention: https://github.com/dotnet/winforms/blob/main/docs/testing.md#naming
```suggestion
    public void StronglyTypedResourceBuilder_Create_NullBaseName_ThrowsArgumentNull()
```

File: src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+        var compileUnit = StronglyTypedResourceBuilder.Create(
+            resxFile: temp.Path,
+            baseName: "Resources",
+            generatedCodeNamespace: "Namespace",
+            s_cSharpProvider,
+            internalClass: false,
+            out _);
+
+        MemoryStream resourceStream = new();
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs:270]:
```suggestion
        using MemoryStream resourceStream = new();
```

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 7]
```diff
@@ -1213,4 +1272,46 @@ Press Ctrl+Enter to accept Text.</value>
...
+  </data>
+  <data name="NonStringPropertyDetailedComment" xml:space="preserve">
+    <value>   Looks up a localized resource of type {0} similar to {1}.</value>
+  </data>
+  <data name="ResMgrPropertyComment" xml:space="preserve">
+    <value>   Returns the cached ResourceManager instance used by this class.</value>
+  </data>
+  <data name="StringPropertyComment" xml:space="preserve">
+    <value>   A strongly-typed resource class, for looking up localized strings, etc.</value>
+  </data>
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:1313]: This looks a dup of 
image

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/ITargetAwareCodeDomProvider.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+using System.ComponentModel;
+
+namespace System.Resources.Tools
+{
+    /// <summary>
+    ///  Defines an interface that enables the strongly typed resource builder (<see cref="StronglyTypedResourceBuilder"/>
+    ///  object) to determine which types and properties are available so it can emit the correct Code Document Object
+    ///  Model (CodeDOM) tree.
+    ///
+    ///  This API supports the product infrastructure and is not intended to be used directly from your code.
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/ITargetAwareCodeDomProvider.cs:14]:
```suggestion
    /// <para>
    ///   Defines an interface that enables the strongly typed resource builder (<see cref="StronglyTypedResourceBuilder"/>
    ///   object) to determine which types and properties are available so it can emit the correct Code Document Object
    ///   Model (CodeDOM) tree.
    /// </para>
    /// <para>
    ///   This API supports the product infrastructure and is not intended to be used directly from your code.
    /// </para>
```

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/ITargetAwareCodeDomProvider.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public interface ITargetAwareCodeDomProvider
+    {
+        /// <summary>
+        ///  Indicates whether the specified type on the project target framework has a specified named property.
+        ///
+        ///  This API supports the product infrastructure and is not intended to be used directly from your code.
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/ITargetAwareCodeDomProvider.cs:22]:
```suggestion
        /// <para>
        ///   Indicates whether the specified type on the project target framework has a specified named property.
        /// </para>
        /// <para>
        ///   This API supports the product infrastructure and is not intended to be used directly from your code.
        /// </para>
```

**Review by RussKie — COMMENTED** [review]
Few other things to consider:
* NRT
* IIUIC, `StronglyTypedResourceBuilder` facilitates resource refereces at compile time (e.g., something we view in VS). Adding a ["verify" test](https://github.com/dotnet/winforms/blob/d799ec671eca9caa51cb050eebbc53754d842725/src/System.Windows.Forms.Analyzers.CSharp/tests/UnitTests/System/Windows/Forms/Generators/ApplicationConfigurationInitializeBuilderTests.cs#L170) would be helpful as well to help eyeball the output.
[This sample](https://docs.microsoft.com/en-us/dotnet/api/system.resources.tools.stronglytypedresourcebuilder?view=netframework-4.8#examples) shows how to get the generated content. 
* Does the ported implementation emit the same output as .NET Framework? If so, may be add a test that would compare the code emitted to a template produced by .NET Framework?

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 10]
```diff
@@ -1,5 +1,64 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
+  <!--
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: This is the first time we edited in a long time and it is going to be rare to modify this file.  This extra `inactive/commented out` section in the resource file is providing no value but increasing the compiled binary size. There was planned task to remove this commented section for every resource file in winforms to reduce binary footprint.  We should make a stand here on why we wanted to add this section back ( I know it is automatically added by Resource editor) and if there is any value that could offset binary foot print increase.

File: src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs:42]: This seems redundant and makes the TestExplorer a pain to use, but I'll change.

File: src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/tests/UnitTests/System/Resources/Tools/StronglyTypedResourceBuilderTests.cs:270]: Can't- the ResourceWriter takes ownership.

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:61]: I don't want to be fighting the tooling here, leaving this in as it has no ill effects and only makes us do additional work to keep removing it if we use the resource editor.

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 7 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: Miscopy, fixed

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 1 above)
Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:61]: Well, we measured the binary size increase due to this additional comments in the past and removed as many as possible. If we are putting this back only because editor gonna add when we edit next time, does not justify the increase in size.

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,949 @@
...
+                    string typeName = node.GetValueTypeName((AssemblyName[])null);
+                    Type type = Type.GetType(typeName);
+                    string valueAsString = node.GetValue((AssemblyName[])null).ToString();
+                    data = new ResourceData(type, valueAsString);
+                }
+                else
+                {
+                    // If the object is null, we don't have a good way of guessing the type. Use Object. This will be
+                    // rare after WinForms gets away from their resource pull model in Whidbey M3.
+                    Type type = (entry.Value == null) ? typeof(object) : entry.Value.GetType();
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:187]:
```suggestion
                    Type type = (entry.Value is null) ? typeof(object) : entry.Value.GetType();
```

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,949 @@
...
+
+            codeCompileUnit.UserData.Add("AllowLateBound", value: false);
+            codeCompileUnit.UserData.Add("RequireVariableDeclaration", value: true);
+
+            CodeNamespace codeNamespace = new(generatedCodeNamespace);
+            codeNamespace.Imports.Add(new CodeNamespaceImport("System"));
+            codeCompileUnit.Namespaces.Add(codeNamespace);
+
+            // Generate class
+            CodeTypeDeclaration classType = new CodeTypeDeclaration(className);
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:370]:
```suggestion
            CodeTypeDeclaration classType = new(className);
```

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,949 @@
...
+
+            string text;
+            if (isString)
+            {
+                text = string.Format(SR.StringPropertyComment, valueAsString);
+            }
+            else
+            {
+                // Stream or Object
+                if (valueAsString == null || string.Equals(typeName, valueAsString))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:771]:
```suggestion
                if (valueAsString is null || string.Equals(typeName, valueAsString))
```

**Review by lonitra — COMMENTED** [review]
nits

**Comment by JeremyKuhne** [comment]
x86 debug tests crashed without a dump :/

**Comment by JeremyKuhne** [comment]
@RussKie 

> NRT

?

> IIUIC, StronglyTypedResourceBuilder facilitates resource refereces at compile time (e.g., something we view in VS). Adding a ["verify" test](https://github.com/dotnet/winforms/blob/d799ec671eca9caa51cb050eebbc53754d842725/src/System.Windows.Forms.Analyzers.CSharp/tests/UnitTests/System/Windows/Forms/Generators/ApplicationConfigurationInitializeBuilderTests.cs#L170) would be helpful as well to help eyeball the output.

I have programmatic tests that validate the output classes and methods are correct via reflection.

> Does the ported implementation emit the same output as .NET Framework? If so, may be add a test that would compare the code emitted to a template produced by .NET Framework?

It is close, but not identical. I don't want to do text comparisons as it doesn't technically matter- this is a .design file. I kept field names the same for now but we may end up changing them to be to our current style based on user feedback. Goal here is to make this low maintenance by having the code up to date with thorough rock-solid tests (no text based comparisons).

**Review by lonitra — APPROVED** [review]
Looks good!

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:61]: This is not about fighting tooling, this is about not committing unnecessary changes. Git allows staging/unstaging hunks.

File: src/System.Design/src/ApiCompatBaseline.netcoreapp.txt [hunk 14]
```diff
@@ -1,323 +0,0 @@
-Compat issues with assembly System.Design:
```
Comment (RussKie) [inline src/System.Design/src/ApiCompatBaseline.netcoreapp.txt:1]: Why was this deleted?

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 10 above)
Comment (RussKie) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: This is unaddressed

**Review by RussKie — COMMENTED** [review]
Why was this merged?

**Comment by RussKie** [comment]
> > NRT
> 
> ?

Newly added code should be nullable-annotated.

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 10 above)
Comment (DustinCampbell) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: FWIW, I'm not sure that I understand the ask. The comment in this RESX file is generated by tooling and is _not_ included in the compiled binary. Is there a misunderstanding about RESX? Note that the dotnet/runtime doesn't remove this comment, so why would WinForms? https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/Resources/Strings.resx

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: Comment of this size in each resx is 3KB extra.  I am trying to recollect if it was just source size increase, given you are confirming it is not compiled into binary. Winforms has so many resx files and in total it was considerable amount.  There was a discussion long back and decided to remove this from resx at that time.  If this is a onetime operation ( and very rare to touch this file again), I still vote to remove it.

Comment (RussKie) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: A commit must only contain necessary changes; it doesn't matter how changes occurred - manually or via some tooling, only required changes should be staged and committed. Especially for files that get changed once in a blue moon.
Staging/unstaging hunks in git is a straight-forward operation.

Comment (DustinCampbell) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:3]: > If this is a onetime operation ( and very rare to touch this file again), I still vote to remove it.

Your suggestion to remove this will require engineering cost to re-remove the comment whenever the file is updated via tooling. So, removing it here will mean that this conversation will happen again later. And, there's no value because comments in the RESX don't get embedded into the compiled binary. ResGen ignores when it builds the `.resources`. If comments _are_ increasing the size of any compiled binary, that is a bug that should be solved in the build tools.
